### PR TITLE
Magicball in VR

### PR DESCRIPTION
**Pull request by agrande**
VR magicball + a bunch of non-VR improvements
To enable the 3D model replacement on desktop, add `ball3d=true` to the url.

**Preview here:** https://pr-624.lba2remake.net/#ball3d=true

### CODE REVIEW

**Review by slwilliams — APPROVED** [review]

File: src/controls/vr/mappings.ts [hunk 1]
```diff
@@ -10,42 +10,164 @@ interface BtnMapping {
...
+                handler: handlePressed
+            };
+        } else if (handedness === 'right') {
+            action = {
+                btn: 'xr-standard-trigger',
+                handler: handlePressed
+            };
+        }
+    }
+    if ('xr-standard-squeeze' in components) {
```
[the remark this answers is not in the document]
Comment (agrande) [reply src/controls/vr/mappings.ts:61]: I'm not sure, having the ball and throwing it is similar to grabbing an object, which usually works with the squeeze button.

File: src/game/loop/animAction.ts [hunk 2]
```diff
@@ -125,10 +125,11 @@ export const THROW = (action: AnimAction, { actor, game, scene }: AnimActionCont
     );
 };
 
-export const THROW_MAGIC = (_action: AnimAction, { actor, game, scene }: AnimActionContext) => {
-    MagicBall.load(game, scene, actor.physics.position).then((mb: MagicBall) => {
-        mb.throw(actor.physics.temp.angle, actor.props.entityIndex);
-    });
+export const THROW_MAGIC = async (_action: AnimAction, ctx: AnimActionContext) => {
+    const { actor, game, scene } = ctx;
+    await MagicBall.instance.init(game, scene);
```
[the remark this answers is not in the document]
Comment (agrande) [reply src/game/loop/animAction.ts:130]: I didn't notice it, thanks for pointing this out, I'll investigate.
There seems to be a delay in general to the update of the ball position, I'll look into that.

File: src/game/loop/hero.ts [hunk 3]
```diff
@@ -293,6 +295,87 @@ function firstPersonPunching(game: Game, scene: Scene, time: Time) {
...
+            TMP_VEC.copy(velocityHistory[i]);
+            TMP_VEC.multiplyScalar(w);
+            WMA.add(TMP_VEC);
+            dta += deltaHistory[i] * w;
+        }
+        WMA.divideScalar(TOTAL_WEIGHTS);
+        dta = dta / TOTAL_WEIGHTS;
+        const strength = WMA.length() / dta;
+        // Throw only if the motion is strong enough
+        if (strength > THROW_MB_THRESHOLD) {
```
[the remark this answers is not in the document]
Comment (agrande) [reply src/game/loop/hero.ts:370]: I'm completely reworking that part in any case, to make the physics more realistic in VR, and I'm getting rid of the threshold altogether, so if you don't move the ball just drops to the floor.

File: src/game/loop/animAction.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (agrande) [reply src/game/loop/animAction.ts]: Fixed that

File: src/controls/vr/mappings.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (agrande) [reply src/controls/vr/mappings.ts:61]: I'll leave it as it is for now, until we discuss controls in VR in a more general way.

**Pull request merged**